kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -1412,6 +1471,70 @@ impl<'tcx> GotocCtx<'tcx> {
         }
         if typ.is_trait() { Some(typ) } else { None }
     }
+
+    /// This function provides an iterator that traverses the data path of a receiver type. I.e.:
+    /// the path that leads to the data pointer.
+    ///
+    /// E.g.: For `Rc<dyn T>` where the Rc definition is:
+    /// ```
+    /// pub struct Rc<T: ?Sized> {
+    ///    ptr: NonNull<RcBox<T>>,
+    ///    phantom: PhantomData<RcBox<T>>,
+    /// }
+    ///
+    /// pub struct NonNull<T: ?Sized> {
+    ///    pointer: *const T,
+    /// }
+    /// ```
+    ///
+    /// The behavior will be:
+    /// ```
+    /// let it = self.receiver_data_path(rc_typ);
+    /// assert_eq!(it.next(), Some(String::from("ptr", non_null_typ));
+    /// assert_eq!(it.next(), Some(String::from("pointer", raw_ptr_typ));
+    /// assert_eq!(it.next(), None);
+    /// ```
+    ///
+    /// Pre-condition: The argument must be a valid receiver for dispatchable trait functions.
```

Comment:
Can we assert this

Reply:
Not specifically. We could create a copy of the iterator, iterate over and ensure that it ends in a fat pointer. But this would pass for other structures that has similar properties, such as a custom struct that just wraps a fat pointer.